schema_enforcer/schemas/jsonschema.py

```diff
@@ -58,7 +58,14 @@ def validate(self, data, strict=False):
         for err in validator.iter_errors(data):
 
             has_error = True
-            self.add_validation_error(err.message, absolute_path=list(err.absolute_path))
+
+            if 'errMessage' in err.schema:
+                message = err.schema['errMessage']
+                message = message.replace("$iData", str(err.instance))
```

Comment:
Nice work! You could even make the `iData` variable dynamic and couple it to defined the data structure so the user can use the property key prepended with `$`:
```suggestion
                
                keyName = '$' + err.absolute_path[-1]
                message = err.schema['errMessage']
                
                if keyName in message:
                    message = message.replace(keyName, str(err.instance))
```
Then then schema would look like:
```
fqdn:
  type: string
  pattern: regex
  errMessage: $fqdn is not valid
ip_address:
  type: string
  format: ipv4
  errMessage: $ip_address is not valid
etc...
```

Reply:
I think you could even do it on one line:

```suggestion
                message = err.schema['errMessage']
                message = message.replace('$' + err.absolute_path[-1], str(err.instance))
```

Reply:
Thanks for the suggestion!
I am only worried about how you index the path. Since:

> The deque can be empty if the error happened at the root of the instance.

Reply:
Oh good catch, I guess the potential `IndexError` could be caught with a `try/except` or similar...

Reply:
Been doing a bit of testing with the following test schema:

```
type: "object"
errMessage: Root error
additionalProperties: false
properties:
  data:
    type: "object"
    properties:
      fqdn:
        type: string
        pattern: regex
        errMessage: "'$fqdn' is not valid. Please see https://confluence.test.com/schema"
      ip_address:
        type: string
        format: ipv4
        errMessage: "'$ip_address' is not valid. Please see https://confluence.test.com/schema"
```
When validating the following data the `IndexError: deque index out of range` error is observed:
```
data:
  fqdn: host
  ip_address: 1.2.3.400
data2:
  fqdn: host
  ip_address: 1.2.3.500
```
The error observed:
```
File "/usr/lib/python3.9/site-packages/schema_enforcer/schemas/jsonschema.py", line 67, in validate
    message = message.replace('$' + err.absolute_path[-1], str(err.instance))
IndexError: deque index out of range
```
One solution that seems to work is to ignore the `errMessage` key at the root level i.e. if `err.absolute_path` evaluates to an empty deque:
```
            # Custom error message handling
            if 'errMessage' in err.schema:

                message = err.schema['errMessage']
                if len(err.absolute_path) > 0:
                    message = message.replace('$' + err.absolute_path[-1], str(err.instance))
                else:
                    message = err.message

            else:
                message = err.message

            self.add_validation_error(message, absolute_path=list(err.absolute_path))
```
Then you get the built-in error messages back from JSON Schema at the root level, and custom error message processing for anything nested:
```
# schema-enforcer validate
FAIL | [ERROR] Additional properties are not allowed ('data2' was unexpected) [FILE] .//test.yml [PROPERTY]
FAIL | [ERROR] 'host' is not valid. Please see https://confluence.test.com/schema [FILE] .//test.yml [PROPERTY] data:fqdn
FAIL | [ERROR] '1.2.3.400' is not valid. Please see https://confluence.test.com/schema [FILE] .//test.yml [PROPERTY] data:ip_address
```

Reply:
Or you could preserve custom error messaging at the root level with an additional `if` statement skipping the `replace` if deque is empty:

```
            message = err.message

            if 'errMessage' in err.schema:
                message = str(err.schema['errMessage'])
                if len(err.absolute_path) > 0:
                    message = message.replace('$' + err.absolute_path[-1], str(err.instance))

            self.add_validation_error(message, absolute_path=list(err.absolute_path))
            ```

Reply:
Happy New Year Everyone!

Just had a thought about the above... If there is a custom message found at the root level, we could check for a reserved variable like $root in the message and use the replace method on that to dump the entire data structure should the user so desire:

            message = err.message

            if 'errMessage' in err.schema:
                message = str(err.schema['errMessage'])
                if len(err.absolute_path) > 0:
                    message = message.replace('$' + err.absolute_path[-1], str(err.instance))
                else:
                    message = message.replace('$root', str(err.instance))

            self.add_validation_error(message, absolute_path=list(err.absolute_path))

It seems a more complete feature to give the user the option to dump the root data structure in a custom error at the root level if they need to.

Reply:
Happy New Year! Sorry for the delay in getting back with you. Thank you so much for opening this!

- I like the idea of using $variable for specific replacement rather than $iData
- I'm having trouble thinking about a case where an error at the root would be more useful than that built-in jsonschema error. I like the idea of evaluating length than letting the built in error bubble up, and just documenting that custom errors at root aren't supported. We could also add a warning message on schema load to ensure the user is aware that a custom error message at root will be skipped.
- We could just print the root error message without replacement and/or add a `$root` keyword. On initial pass, I don't like this idea for two reasons: 1. The behavior is applied differentially depending on where the error message exists. We replace sometimes but not other times. 2. The `$root` is a bit of magic, it's not quite intuitive what's going on there, where you can use `$root` and where you cant. Both of those reasons make me think it'd introduce complexity to the user in a way that doesn't seem worth the trade off of adding this functionality to me.

Reply:
Hey Phil, thanks for coming back!

I understand what you mean about the root error functionality, it would introduce complexity to the user and the feature would be cleaner without it. I've removed it from the examples below.

However there are a few edge cases that I've identified with using a specific $variable. It causes a few issues when a property is an array/list or when a sub-schema definition is referenced. Firstly, here is an example of the issue when using an array/list:

Example schema:
```
type: object
additionalProperties: false
properties:
  data:
    type: object
    properties:
      name:
        type: string
        pattern: regex
        err_message: "'$name' is not valid. Please see docs."
      aliases:
        type: array
        items:
          type: string
          pattern: regex
          err_message: "'$item' is not valid. Please see docs."
```
If we then take the following example data:
```
data:
  name: host
  aliases:
    - alias-1
    - alias-2
```
The errors returned are:
```
FAIL | [ERROR] 'host' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:name
FAIL | [ERROR] '$item' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:aliases:0
FAIL | [ERROR] '$item' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:aliases:1
```
As can be seen above, the $item var is ignored because there is no logic for it. The property itself is a numerical index of the list. I experimented overcoming this by checking if the last item in the deque is an integer, and if so, look for and replace the $item var:
```
            has_error = True
            err_message = err.message

            if 'err_message' in err.schema:
                err_message = str(err.schema['err_message'])
                if len(err.absolute_path) > 0:
                    if type(err.absolute_path[-1]) is int:
                        err_message = err_message.replace('$item', str(err.instance))
                    else:
                        err_message = err_message.replace('$' + str(err.absolute_path[-1]), str(err.instance))

            self.add_validation_error(err_message, absolute_path=list(err.absolute_path))
```
This solution works for lists and returns errors correctly:
```
FAIL | [ERROR] 'host' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:name
FAIL | [ERROR] 'alias-1' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:aliases:0
FAIL | [ERROR] 'alias-2' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:aliases:1
```
However, the next issue kind of makes the above redundant, and it's when a sub schema is referenced. Here is the example schema:
```
type: object
additionalProperties: false
properties:
  data:
    type: object
    properties:
      name:
        $ref: "#/$defs/regex_sub_schema"
      aliases:
        type: array
        items:
          $ref: "#/$defs/regex_sub_schema"

$defs:
  regex_sub_schema:
    type: string
    pattern: regex
    err_message: "'$name' is not valid. Please see docs."
```
The custom error message moves to the sub schema and the property is inherited from whatever references it. Using the same test data, here are the errors returned:
```
FAIL | [ERROR] 'host' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:name
FAIL | [ERROR] '$name' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:aliases:0
FAIL | [ERROR] '$name' is not valid. Please see docs. [FILE] .//test.yml [PROPERTY] data:aliases:1
```
As we can see above, it works fine for $name because that's the correct variable for the first property, however subsequent properties won't work. Unfortunately you can't have the custom message alongside the reference:
```
[...]
      name:
        $ref: "#/$defs/regex_sub_schema"
        err_message: "'$name' is not valid. Please see docs."    <<<< This doesn't work.
      aliases:
        type: array
        items:
          $ref: "#/$defs/regex_sub_schema"
          err_message: "'$item' is not valid. Please see docs."

$defs:
  regex_sub_schema:
    type: string
    pattern: regex
```
This brings us back to using a common var for all properties. In this PR it's been dubbed $iData (the 'i' being an abbreviation of 'instance') but it could be changed to just $instance. This is a little more intuitive and line with JSON Schema terminology and under-the-hood variable naming:
```
            has_error = True
            err_message = err.message

            # Custom error message handling
            if len(err.absolute_path) > 0 and 'err_message' in err.schema:
                err_message = str(err.schema['err_message'])
                err_message = err_message.replace('$instance', str(err.instance))

            self.add_validation_error(err_message, absolute_path=list(err.absolute_path))
```
In the patch above, we check:
- we are not at root level
- a custom error message exists

If the above conditions evaluate to true, then override the error message and replace any occurrences of $instance with the actual JSON Schema instance:
```
FAIL | [ERROR] 'host' is not valid. Please see docs. host [FILE] .//test.yml [PROPERTY] data:name
FAIL | [ERROR] 'alias-1' is not valid. Please see docs. alias-1 [FILE] .//test.yml [PROPERTY] data:aliases:0
FAIL | [ERROR] 'alias-2' is not valid. Please see docs. alias-2 [FILE] .//test.yml [PROPERTY] data:aliases:1
```
If `err_message` is defined at the root level, it is simply ignored. With this patch the example schema's above would look like:
```
[...]
      name:
        type: string
        pattern: regex
        err_message: "'$instance' is not valid. Please see docs."
      aliases:
        type: array
        items:
          type: string
          pattern: regex
          err_message: "'$instance' is not valid. Please see docs."
[...]
```
```
[...]
      name:
        $ref: "#/$defs/regex_sub_schema"
      aliases:
        type: array
        items:
          $ref: "#/$defs/regex_sub_schema"

$defs:
  regex_sub_schema:
    type: string
    pattern: regex
    err_message: "'$instance' is not valid. Please see docs."
```
There are less moving parts for the user, `err_message` and `$instance` is all they need to construct a meaningful message that satisfies their custom requirement.

Apologies for the long post! I really like Schema Enforcer and think this would be such a great feature for users if it can be implemented in the right way.

Reply:
Thanks for doing all this testing, really appreciate the diligence here.

Given the complexity of using `$variable`, I think it makes a ton of sense to use the `$instance` nomenclature/methodology. Intuitive for the user and easier to maintain for the developer.

Reply:
Are there any edge cases that the `$instance` solution you've proposed doesn't catch correctly (besides a message existing at root)?

Reply:
Hey Phil, no problem!

So I've tried quite a few schemas in my local testing and using `$instance` does indeed seem to work fine in every scenario I've looked at. One thing to bear in mind is that it will only return the parent error (this is the default behaviour of Schema Enforcer anyway). When the parent error is the result of sub-errors when using schema composition keywords like `anyOf`, the `ValidationError.context` attribute can be used to return the sub-errors. But returning these sub-errors in SE is crossing over into a whole different feature request.

Reply:
Acknowledged, thanks for all the testing.

It would be awesome to get those tests into unit tests to validate future changes don't break things, and that different custom message cases work as we expect them to.

Reply:
Hi Phill,

I would like to open a new PR for this feature as @anakhalil has moved on to a new adventure and likely won't have the time to flesh out the unit/integration tests required.

Hopefully this will be fairly soon!

Reply:
Sounds good, thank you! :)

Reply:
I've pushed some commits to this PR that introduce some unit tests for custom error message handling and also a README snippet. There were a few hurdles with the unit tests, for some reason I could not get it to work with the schema broken up into separate files, so had to adjust it so it was all in one file. Even then it would not allow me to use references to definitions from within the same file.

However the tests do pass. I tried to make the tests in line with the existing ones and not interfere with them so there is some repeated code. Hopefully it's a good starting point anyway!